### DISCUSSION TITLE
Support for feather files

### DISCUSSION THREAD

**Discussion by aftersought** [duckdb/duckdb#3729] [Ideas]
You currently support parquet files in queries, e.g. `DESCRIBE SELECT * FROM 'test.parquet';` but I don't see similar support for feather files despite your engine being integrated with Arrow by now. Is this planned for future release?

This would make DuckDB a very handy tool for quick command-line analysis of feather files.

**Comment by hannes** [comment]
This is unlikely to be added to DuckDB core because it introduce a hard arrow dependency at runtime. Arrow itself has multiple transitive dependencies. If possible to link it all statically, we could consider creating an extension.

**Reply by aftersought** [reply]
I cannot comment on the policy of not having Arrow as dependency, but I would consider Feather to be a stable and somewhat popular format. In our use case (and I suspect more generally) it offers faster access at the cost of a larger disk footprint, a very acceptable trade-off for cases where speed matters. Given that DuckDB in general positions itself as a "faster" analytics engine compared to standard alternatives (say, pandas) I would think it is logical to support "faster" formats.

**Reply by hu6360567** [reply]
Is it possible to implement feather file import by plugin?

**Comment by hannes** [comment]
What works really well already is to have Arrow open the feather file and then query that from DuckDB. Even does pushdown and streaming etc. Maybe @Alex-Monahan can whip up an example?

**Reply by aftersought** [reply]
We have played with / benchmarked DuckDB + Arrow over feather and parquet files using that functionality, so I believe I am familiar with it. Unfortunately that does not help with quick&dirty analysis using CLI, something where DuckDB would shine in comparison with pandas (which necessitates opening python console + writing boilerplate commands first).

**Comment by wisefool769** [comment]
I'm also a fan of this feature fwiw. 
DuckDB is amazing, and I want to see its performance get as close as possible to kdb's -- which will require it to support more read-efficient file formats than parquet (as well as more flexible partitioning schemes). Supporting feather would be a great step in that direction!

**Reply by ssh352** [reply]
are you suggesting that feature is more read efficient than parquet? @wisefool769

**Reply by mrietveld** [reply]
@ssh352 Feather is can be read _faster_ than parquet because parquet emphasizes compression more then Feather. Feather is also of course "closer" to the arrow in-memory format, which also decreases read (and write) times. (AFAIK, feather is basically arrow's serialization method for it's in-memory format, and also what's used in, e.g., `pyarrow.ipc.open_file`). 

See https://stackoverflow.com/a/48097717/376390, from Wes McKinney, co-creator of Arrow. In the same answer, there is a ("anecdotal") benchmark showing the speeds of feather/arrow vs parquet. 

Not sure if you meant faster by "efficient" or something else. 

However, I'm not totally sure that _querying_ feather/Arrow IPC will always be faster, since I have the vague feeling that there are mechanisms in parquet that allow faster querying..  😬

**Reply by torchss** [reply]
@mrietveld Would supporting Feather allow "zero copy" behavior when loading them into DuckDB vs parquet? If so, will using compression on Feather affect this "zero copy" behavior?

**Reply by mrietveld** [reply]
@torchss So, if I understand correctly, DuckDB uses Arrow internally. When feather is read in, it's read into the Arrow format, so a zero-copy is possible, but that depends on whether DuckDB does that, obviously. If you apply compression to Feather, that would slow down reading in Feather from disk, but it's still read in to the Arrow in-memory format, since compression (usually) only applies to the serialized format.

**Reply by Tishj** [reply]
DuckDB does not use Arrow internally

**Reply by torchss** [reply]
@mrietveld While DuckDB does not use Arrow internally, I believe there are many options to get efficient semantics - Flight, reading the Arrow outside DuckDB, getting a Substrait plan from DuckDB and running the compute in Arrow etc - each has its own pros and cons, so I am trying to see which approach has the least amount of moving parts vs reliability ratio.

**Comment by paultiq** [comment]
FYI, while it's not a native duckdb solution, is Python specific, and is certainly a work-in-progress, I hope this is a helpful: 

I created [fabduckdb](https://github.com/paultiq/fabduckdb) which allows you to insert functions that return table-like* or file* objects into [duckdbpyconnections](https://duckdb.org/docs/api/python/reference/#duckdb.DuckDBPyConnection). This allows you to add a feather reader. 

Here's two examples of how you could do that:

#### Using PyArrow to read the feather:
In this version, readftr() creates a PyArrow Table which is registered in the connection. The query is rewritten to use the registered name of the Table. 
```
# Feather (returns a PyArrow Table)

import fabduckdb
import duckdb
import pyarrow as pa

## Sample Data
import pandas as pd
irisdf = pd.read_csv('https://raw.githubusercontent.com/mwaskom/seaborn-data/master/iris.csv')
filename = "my.ftr"
irisdf.to_feather(filename)

## Registration
fabduckdb.register_function(
        "readftr",
        lambda input, con: pa.feather.read_table(input),  # or, using pandas: pd.read_feather(input)
        generates_filepath=False,
    )

## Usage
df = duckdb.connect().execute("select species, min(sepal_length), max(sepal_length) from readftr('my.ftr') group by species").df()
print(df)
```

#### Using PyArrow to Read a Feather and Transcode to Parquet
In this version, readftr_asparquet() transcodes the feather to parquet, and rewrites the query so DuckDB is querying the parquet file. 

```
import pandas as pd
import fabduckdb
import duckdb
import pyarrow as pa
import pyarrow.parquet as pq

## Sample Data
irisdf = pd.read_csv('https://raw.githubusercontent.com/mwaskom/seaborn-data/master/iris.csv')
filename = "my.ftr"
irisdf.to_feather(filename)

## Registration
fabduckdb.register_function(
        "readftr_asparquet",
        lambda input, filename, con: pa.parquet.write_table(pa.feather.read_table(input), filename),
        generates_filepath=True,
    )

## Usage: 
df = duckdb.connect().execute("select species, min(sepal_length), max(sepal_length) from readftr_asparquet('my.ftr') group by species").df()
print(df)
```
#### Notes

\* Table-like: any supported input type by `duckdb.register()`: DataFrames, Arrow Tables, etc. Files: any filename (parquet & csv) that can be queried as: `select * from 'filename.ext'`
\** duckdbpyrelations are not supported
\*** Caution: This obviously is a very new project, so ymmv. Please file any issues you find.

**Comment by Tishj** [comment]
Through the `arrow` community extension (https://duckdb.org/community_extensions/extensions/arrow.html) we support the IPC format, according to the [official docs](https://arrow.apache.org/docs/python/feather.html)

> Version 2 (V2), the default version, which is exactly represented as the Arrow IPC file format on disk. V2 files support storing all Arrow data types as well as compression with LZ4 or ZSTD. V2 was first made available in Apache Arrow 0.17.0.

Which sounds to me like it will directly support reading Feather (v2) files?

**Reply by torchss** [reply]
It seems like the arrow community extension works only in DuckDB v1.3.0 and misleadingly loads but fails to work on any engines below that. This issue is documented but cannot be currently discovered by an LLM/code assistant or search engine as the details are behind a very broad and ambiguous title:  https://github.com/paleolimbot/duckdb-nanoarrow/issues/26

I pointed this out at https://github.com/paleolimbot/duckdb-nanoarrow/issues/26#issuecomment-2951576430

So yes, DuckDB v1.3.0, but nothing below it, will open feather v2.
